sql/core/src/main/scala/org/apache/spark/sql/execution/window/RankLimitExec.scala

```diff
@@ -0,0 +1,280 @@
+import org.apache.spark.sql.execution.metric.SQLMetrics
+import org.apache.spark.util.collection.Utils
+
+
+sealed trait RankLimitMode
+
+case object Partial extends RankLimitMode
+
+case object Final extends RankLimitMode
+
+
+
+/**
+ * This operator is designed to filter out unnecessary rows before WindowExec,
+ * for top-k computation.
+ * @param partitionSpec Should be the same as [[WindowExec#partitionSpec]]
+ * @param orderSpec Should be the same as [[WindowExec#orderSpec]]
+ * @param rankFunction The function to compute row rank, should be RowNumber/Rank/DenseRank.
+ */
+case class RankLimitExec(
+    partitionSpec: Seq[Expression],
+    orderSpec: Seq[SortOrder],
+    rankFunction: Expression,
+    limit: Int,
+    mode: RankLimitMode,
+    child: SparkPlan) extends UnaryExecNode {
+  assert(orderSpec.nonEmpty && limit > 0)
+
+  private val shouldPass = child match {
+    case r: RankLimitExec =>
+      partitionSpec.size == r.partitionSpec.size &&
+        partitionSpec.zip(r.partitionSpec).forall(p => p._1.semanticEquals(p._2)) &&
+        orderSpec.size == r.orderSpec.size &&
+        orderSpec.zip(r.orderSpec).forall(o => o._1.semanticEquals(o._2)) &&
+        rankFunction.semanticEquals(r.rankFunction) &&
+        mode == Final && r.mode == Partial && limit == r.limit
+    case _ => false
+  }
+
+  private val shouldApplyTakeOrdered: Boolean = {
+    rankFunction match {
+      case _: RowNumber => limit < conf.topKSortFallbackThreshold
+      case _: Rank => false
+      case _: DenseRank => false
+      case f => throw new IllegalArgumentException(s"Unsupported rank function: $f")
+    }
+  }
+
+  override def output: Seq[Attribute] = child.output
+
+  override def requiredChildOrdering: Seq[Seq[SortOrder]] = {
+    if (shouldApplyTakeOrdered) {
+      Seq(partitionSpec.map(SortOrder(_, Ascending)))
+    } else {
+      // Should be the same as [[WindowExec#requiredChildOrdering]]
+      Seq(partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec)
+    }
+  }
+
+  override def outputOrdering: Seq[SortOrder] = {
+    partitionSpec.map(SortOrder(_, Ascending)) ++ orderSpec
+  }
+
+  override def requiredChildDistribution: Seq[Distribution] = mode match {
+    case Partial => super.requiredChildDistribution
+    case Final =>
+      // Should be the same as [[WindowExec#requiredChildDistribution]]
+      if (partitionSpec.isEmpty) {
+        AllTuples :: Nil
+      } else ClusteredDistribution(partitionSpec) :: Nil
+  }
+
+  override def outputPartitioning: Partitioning = child.outputPartitioning
+
+  override lazy val metrics = Map(
+    "numOutputRows" -> SQLMetrics.createMetric(sparkContext, "number of output rows"))
+
+  private lazy val ordering = GenerateOrdering.generate(orderSpec, output)
+
+  private lazy val limitFunction = rankFunction match {
+    case _: RowNumber if shouldApplyTakeOrdered =>
+      (stream: Iterator[InternalRow]) =>
+        Utils.takeOrdered(stream.map(_.copy()), limit)(ordering)
```

Comment:
By default, `spark.sql.execution.topKSortFallbackThreshold` is set to a pretty big number (Integer.MAX_VALUE - 15). Therefore, by default anyway, this line of code will attempt to do a sort with guava regardless of the size of the rank limit.

For example, using your example [here](https://github.com/apache/spark/pull/34367#issuecomment-949516811), but changing the where clause to `col("rank") <= 1000000`, I get:

```
java.lang.OutOfMemoryError: GC overhead limit exceeded
```

Whereas I don't get that with `spark.sql.rankLimit.enabled=false`.

If I set `spark.sql.execution.topKSortFallbackThreshold=10000`, it succeeds with `spark.sql.rankLimit.enabled=true`.

Maybe it needs its own threshold? Or something to make users more aware (since it is not always super clear from the stack traces what was causing the OOMs).

Reply:
```
val TOP_K_SORT_FALLBACK_THRESHOLD =
    buildConf("spark.sql.execution.topKSortFallbackThreshold")
      .doc("In SQL queries with a SORT followed by a LIMIT like " +
          "'SELECT x FROM t ORDER BY y LIMIT m', if m is under this threshold, do a top-K sort" +
          " in memory, otherwise do a global sort which spills to disk if necessary.")
      .version("2.4.0")
      .intConf
      .createWithDefault(ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH)
```

I think we can still share the same threshold. The problem is that its default value is too high, which should be set to a relative smaller number in practice, such as 10000.